File: src/core/retrieval/hybrid.ts
/**
 * 混合检索引擎 - RRF融合
 */

import { BM25Retriever, getBM25Retriever, initBM25Index, clearBM25Index } from '@/core/retrieval/bm25';
import { VectorRetriever, getVectorRetriever, initVectorIndex, clearVectorIndex } from '@/core/retrieval/vector';
import { DomainConfig } from '@/core/domain-config';

export interface RetrievalResult {
    id: string;
    score: number;
    title: string;
    content: string;
    source?: string;
    category?: string;
    bm25Rank?: number;
    vectorRank?: number;
}

function rrfFusion(
    bm25Results: Array<{ id: string; score: number }>,
    vectorResults: Array<{ id: string; score: number }>,
    bm25Weight: number = 0.5,
    vectorWeight: number = 0.5,
    k: number = 60
): Map<string, number> {
    const scores = new Map<string, number>();

    bm25Results.forEach((result, rank) => {
        const score = bm25Weight * (1 / (k + rank + 1));
        scores.set(result.id, (scores.get(result.id) || 0) + score);
    });

    vectorResults.forEach((result, rank) => {
        const score = vectorWeight * (1 / (k + rank + 1));
        scores.set(result.id, (scores.get(result.id) || 0) + score);
    });

    return scores;
}

export class HybridRetriever {
    private bm25: BM25Retriever;
    private vector: VectorRetriever;
    private config: DomainConfig['retrieval'];

    constructor(domain: string, config: DomainConfig['retrieval']) {
        this.bm25 = getBM25Retriever(domain);
        this.vector = getVectorRetriever(domain);
        this.config = config;
    }

    search(query: string, topK?: number): RetrievalResult[] {
        const k = topK || this.config.top_k || 5;
        const searchK = k * 3;

        const bm25Results = this.bm25.search(query, searchK);
        const vectorResults = this.vector.search(query, searchK);

        if (this.config.strategy === 'bm25') {
            return bm25Results.slice(0, k).map(r => ({
                id: r.document.id,
                score: r.score,
                title: r.document.title,
                content: r.document.content,
                source: r.document.source,
                category: r.document.category,
            }));
        }

        if (this.config.strategy === 'vector') {
            return vectorResults.slice(0, k).map(r => ({
                id: r.document.id,
                score: r.score,
                title: r.document.title,
                content: r.document.content,
                source: r.document.source,
                category: r.document.category,
            }));
        }

        const fusedScores = rrfFusion(
            bm25Results.map(r => ({ id: r.id, score: r.score })),
            vectorResults.map(r => ({ id: r.id, score: r.score })),
            this.config.bm25_weight,
            this.config.vector_weight,
            this.config.rrf_k || 60
        );

        const sortedIds = Array.from(fusedScores.entries())
            .sort((a, b) => b[1] - a[1])
            .slice(0, k);

        const allDocs = new Map<string, RetrievalResult>();

        for (const r of bm25Results) {
            allDocs.set(r.id, {
                id: r.document.id,
                score: 0,
                title: r.document.title,
                content: r.document.content,
                source: r.document.source,
                category: r.document.category,
                bm25Rank: bm25Results.indexOf(r) + 1,
            });
        }

        for (const r of vectorResults) {
            const existing = allDocs.get(r.id);
            if (existing) {
                existing.vectorRank = vectorResults.indexOf(r) + 1;
            } else {
                allDocs.set(r.id, {
                    id: r.document.id,
                    score: 0,
                    title: r.document.title,
                    content: r.document.content,
                    source: r.document.source,
                    category: r.document.category,
                    vectorRank: vectorResults.indexOf(r) + 1,
                });
            }
        }

        return sortedIds.map(([id, score]) => ({
            ...allDocs.get(id)!,
            score,
        })).filter(Boolean);
    }
}

const hybridRetrieverCache = new Map<string, HybridRetriever>();

export function getHybridRetriever(domain: string, config: DomainConfig['retrieval']): HybridRetriever {
    if (!hybridRetrieverCache.has(domain)) {
        hybridRetrieverCache.set(domain, new HybridRetriever(domain, config));
    }
    return hybridRetrieverCache.get(domain)!;
}

export async function initAllIndices(domain: string): Promise<void> {
    await Promise.all([
        initBM25Index(domain),
        initVectorIndex(domain),
    ]);
    console.log(`[Retrieval] 领域 ${domain} 所有索引初始化完成`);
}

export async function search(
    query: string,
    domainConfig: DomainConfig,
    topK?: number
): Promise<RetrievalResult[]> {
    const domain = domainConfig.domain.id;
    const retriever = getHybridRetriever(domain, domainConfig.retrieval);
    return retriever.search(query, topK);
}

export async function reinitializeDomainIndex(domain: string): Promise<void> {
    clearBM25Index(domain);
    clearVectorIndex(domain);
    hybridRetrieverCache.delete(domain);

    await initAllIndices(domain);
    console.log(`[Retrieval] 领域 ${domain} 索引已重新初始化`);
}

File: server/core/retrieval/hybrid.py
"""
Hybrid retrieval combining BM25 + Vector with RRF.
"""
from typing import List, Dict, Any, Optional

from core.retrieval.bm25 import BM25Retriever
from core.retrieval.vector import VectorRetriever


class HybridRetriever:
    def __init__(self, bm25_weight: float = 0.6, vector_weight: float = 0.4, rrf_k: int = 60):
        self.bm25 = BM25Retriever()
        self.vector = VectorRetriever()
        self.bm25_weight = bm25_weight
        self.vector_weight = vector_weight
        self.rrf_k = rrf_k

    def index(self, documents: List[Dict[str, Any]]):
        self.bm25.index(documents)
        self.vector.index(documents)

    def search(self, query: str, top_k: int = 5) -> List[Dict[str, Any]]:
        if not self.bm25.documents and not self.vector.documents:
            return []
        fetch_k = min(top_k * 3, 20)
        bm25_results = self.bm25.search(query, top_k=fetch_k)
        vector_results = self.vector.search(query, top_k=fetch_k)

        doc_scores: Dict[str, float] = {}
        doc_data: Dict[str, Dict[str, Any]] = {}

        for rank, result in enumerate(bm25_results):
            key = result["title"]
            rrf_score = self.bm25_weight / (self.rrf_k + rank + 1)
            doc_scores[key] = doc_scores.get(key, 0) + rrf_score
            doc_data[key] = result

        for rank, result in enumerate(vector_results):
            key = result["title"]
            rrf_score = self.vector_weight / (self.rrf_k + rank + 1)
            doc_scores[key] = doc_scores.get(key, 0) + rrf_score
            if key not in doc_data:
                doc_data[key] = result

        sorted_docs = sorted(doc_scores.items(), key=lambda x: x[1], reverse=True)
        results = []
        for key, score in sorted_docs[:top_k]:
            result = doc_data[key].copy()
            result["score"] = score
            result["source"] = "hybrid"
            results.append(result)
        return results